README
autobox version 3.0.1
====================

The autobox pragma allows methods to be called on integers, floats, strings, arrays, hashes, and code references in exactly the same manner as blessed references.

Autoboxing is transparent: values are not blessed into their (user-defined) implementation class (unless the method elects to bestow such a blessing) - they simply use its methods as though they are.

autobox is lexically scoped, and bindings for an outer scope can be overridden or countermanded in a nested scope.

See the POD for more details.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  Scope::Guard

COPYRIGHT AND LICENCE

Copyright (c) 2003-2021 by chocolateboy.

This is free software; you can redistribute it and/or modify it under the
terms of the Artistic License 2.0.


Changes
Revision history for Perl extension autobox

3.0.2 TBC

    - GH #15: fix debugger test (thanks, djerius)
    - upgrade ppport.h from 3.42 to 3.68
    - GH #12: fix doc typo (thanks, guimard)
    - GH #14: fix changelog typo (thanks, EdwardBetts)

3.0.1 Sat 5 May 2018

    - GH #11: fix version declaration on 5.8 (thanks, Grinnz)

3.0.0 Sat 5 May 2018

    - breaking change:
        - the behaviour of UNIVERSAL methods like $native->can and
          $native->isa is now defined as being the same as when autobox
          is not enabled rather than "undefined". (technically, this
          still falls under the rubric of "undefined", but the switch
          from "don't know" to "don't" could break buggy code, so bump
          for safety)
        - add DOES to the list of non-autoboxed methods
    - switch to SemVer i.e. 2.86 (v2.860.0) -> 3.0.0 (v3.0.0)
    - upgrade ppport.h from 3.35 -> 3.42

2.86 Fri 20 Apr 2018

    - GH #9: fix bug which prevented autoboxing working
      under the debugger on perl 5.22+ (thanks, skington)
        - added t/debugger.t
    - GH #8: fix bug which prevented bareword method-calls being
      exempted when the method is a variable e.g. Foo->$bar
    - add operator-overloading note to the gotchas section (GH #7)

2.85 Mon 27 Feb 2017

    - fix failing test under 5.25.10 with -Ddefault_inc_excludes_dot
      (thanks, Kent Fredric)

2.84 Fri 26 Aug 2016

    - compatibility fix for perl >= 5.25
      (thanks, mat813, karenetheridge and eserte)
    - upgrade ppport.h from 3.20 to 3.35

2.83 Sun Feb 1 2015

    - RT #100247: fix assertion failures on 5.21.x perls
      with -DDEBUGGING (thanks, ilmari and Father Chrysostomos)
    - RT #100717: don't hide autobox::universal from PAUSE
      (thanks, ppisar)
    - RT #89754: INSTALLDIRS fix (thanks, Kent Fredric)

2.82 Sat Oct 26 2013

    - simplify test to avoid portability woes

2.81 Sat Oct 26 2013

    - fix failing test on Windows

2.80 Fri Oct 25 2013

    - RT #71777: fix segfault in destructor called during
      global destruction (thanks, Tomas Doran)
        - added t/rt_71777.t
    - fix doc typo (thanks, David Steinbrunner)

2.79 Tue Apr 30 2013

    - allow import arguments to be passed as a hashref
        - added t/import_hashref.t
    - doc tweaks

2.78 Tue Apr 30 2013

    - RT #80400: fix segfault in destructor called in END block
      (thanks, Tokuhiro Matsuno)
        - added t/rt_80400.t

2.77  Thu Dec 13 2012

    - doc tweaks
    - add multiple-arg auto-ref tests

2.76  Wed Nov 21 2012

    - fix breaking tests in perl >= 5.17.5: update error
      message pattern (thanks, rjbs)
    - upgrade ppport.h from 3.19 to 3.20

2.75  Thu Jul 21 2011

    - POD spelling fixes (thanks, Jonathan Yu and gregor herrmann)

2.74  Wed Jul 20 2011

    - portability fix for perl >= 5.14 (thanks, chorny)

2.73  Sun Mar 13 2011

    - Makefile.PL fix

2.72  Fri Jan 28 2011

    - fix conflict with `use re "taint"` (thanks, Peter Rabbitson)

2.71  Thu Sep 23 2010

    - fix for recent perls: remove cargo-cultism

2.70  Wed Mar 17 2010

    - replace autobox_can and autobox_isa with autobox_class.
      this also fixes import, unimport and VERSION
        - added t/version.t
        - renamed t/universal.t -> t/autobox_class.t

2.60  Wed Mar 17 2010

    - fix RT #46814 (thanks, Tye McQueen)
        - added t/rt_46814.t
    - fix RT #49273 (thanks, Daniel Austin)
    - fix RT #55565 (thanks, Schwern)
    - fix RT #55652 (thanks, Schwern)
        - $native->isa and $native->can must now be called as
          $native->autobox_isa and $native->autobox_can
        - added t/rt_55652.t

2.55  Sun May 25 2008

    - fix MANIFEST again - restore Changes

2.54  Sun May 25 2008

    - fix MANIFEST

2.53  Sat May 24 2008

    - add support for UNIVERSAL virtual type
        - added t/universal.t
    - moved autobox::type method to autobox::universal::type subroutine
        - added export.t
    - added t/default.t
    - portability fix for non-gcc compilers (thanks, chris)
    - misc code/documentation fixes/cleanups

2.52  Tue May 20 2008

    - more type fixes

2.51  Tue May 20 2008

    - fix type identification for former INTEGERs and FLOATs
      (thanks, Mitchell N Charity)
        - added type.t
    - fix for perl 5.11 (thanks, Andreas Koenig)
    - document `eval EXPR` gotcha

2.50  Mon May 19 2008

    - add support for INTEGER, FLOAT, NUMBER and STRING
        - added scalar.t
    - updated documentation

2.43  Thu May 15 2008

    - fix @isa bug
        - added t/isa.t
    - scope cleanup
    - documentation tweak

2.42  Tue May 13 2008

    - upgrade ppport.h to 3.13_03 to s/workaround/fix/

2.41  Tue May 13 2008

    - work around $value->$method segfault with non-string method
      names under perls <= 5.8.8
    - added license info

2.40  Mon May 12 2008

    - support @array and %hash (thanks, Yuval Kogman (nothingmuch)
      and Matthijs van Duin (xmath))
        - added t/autoref.t
    - fix $value->$method segfault with undef, integer, float etc.
      (i.e. non-string) method names (thanks, John Goulah)

2.30  Fri May  9 2008

    - support $value->$method, where $method is a method name or
      subroutine reference:
        - added t/name.t
        - added t/coderef.t

2.23  Sun Feb 24 2008

    - rm redundant $^H hacking

2.22  Sun Feb 24 2008

    - added hints.t

2.21  Fri Feb 22 2008

    - merge unimport.t and time_travel.t into unmerge.t
    - more tests

2.20  Thu Feb 21 2008

    - Fix broken merging
        - corrected merge.t
        - added time_travel.t to verify correctness

2.11  Wed Feb 20 2008

    - Windows portability fix: ANSIfy C99-ism (thanks, Taro Nishino)
    - revert broken micro-optimization

2.10  Wed Feb 20 2008

    - fix + add tests for:
        - unimport
        - default namespace(s) in an array ref

2.02  Sun Feb 17 2008

    - doc tweak
    - POD formatting

2.01  Sun Feb 17 2008

    - documentation fix:

t/debugger.pl
#!/usr/bin/env perl

use strict;
use warnings;
use blib;

use autobox { DEFAULT => __PACKAGE__ };

# helper for t/debugger.t
#
# this prints "foo -> bar -> baz -> quux" on perl < 5.22
# and fails with the following error on newer perls:
#
#     Can't locate object method "baz" via package "foo -> bar"
#
# https://github.com/scrottie/autobox-Core/issues/34

sub bar {
    return "$_[0] -> bar";
}

sub baz {
    return "$_[0] -> baz";
}

sub quux {
    print "$_[0] -> quux", $/;
}

'foo'->bar->baz->quux;


t/rt_71777.pl
#!/usr/bin/env perl

use strict;
use warnings;
use blib;

# simplified version of the test case provided by Tomas Doran (t0m)
# https://rt.cpan.org/Ticket/Display.html?id=71777

{
    package Foo;
    use autobox;
    sub DESTROY {
        # silence this distracting (but expected) warning:
        # "(in cleanup) Can't call method "bar" on an undefined value at t/rt_71777.pl line 17 during global destruction."
        no warnings qw(misc);
        $_[0]->{bar}->bar
    }
}

{
    package Bar;
    sub bar { }
}

my $bar = bless {}, 'Bar';
my $foo = bless {}, 'Foo';

$foo->{bar} = $bar;
$bar->{foo} = $foo;


t/autoboxtest.pm
package autoboxtest;

use strict;
use warnings;

our $VERSION = '1.23';
our $value = '';

sub import {
	$VERSION = '3.14';
}

sub unimport {
	$value = 'unimport';
}

sub test {
	return $value;
}

1;


ptable.h
/*
 * This is copied more or less verbatim from the pointer table implementation in sv.c
 */

#include "ppport.h"

#define PTABLE_HASH(ptr) ((PTR2UV(ptr) >> 3) ^ (PTR2UV(ptr) >> (3 + 7)) ^ (PTR2UV(ptr) >> (3 + 17)))

struct PTABLE_entry {
    struct PTABLE_entry     *next;
    void                    *key;
    void                    *value;
};

struct PTABLE {
    struct PTABLE_entry     **tbl_ary;
    UV                      tbl_max;
    UV                      tbl_items;
};

typedef struct PTABLE_entry PTABLE_ENTRY_t;
typedef struct PTABLE            PTABLE_t;

static PTABLE_t * PTABLE_new(void);
static PTABLE_ENTRY_t * PTABLE_find(PTABLE_t *tbl, const void *key);
static void * PTABLE_fetch(PTABLE_t *tbl, const void *key);
static void PTABLE_store(PTABLE_t *tbl, void *key, void *value);
static void PTABLE_grow(PTABLE_t *tbl);
static void PTABLE_clear(PTABLE_t *tbl);
static void PTABLE_free(PTABLE_t *tbl);

/* create a new pointer => pointer table */

static PTABLE_t *
PTABLE_new(void)
{
    PTABLE_t *tbl;
    Newxz(tbl, 1, PTABLE_t);
    tbl->tbl_max = 511;
    tbl->tbl_items = 0;
    Newxz(tbl->tbl_ary, tbl->tbl_max + 1, PTABLE_ENTRY_t*);
    return tbl;
}

/* map an existing pointer using a table */

static PTABLE_ENTRY_t *
PTABLE_find(PTABLE_t *tbl, const void *key) {
    PTABLE_ENTRY_t *tblent;
    const UV hash = PTABLE_HASH(key);
    tblent = tbl->tbl_ary[hash & tbl->tbl_max];
    for (; tblent; tblent = tblent->next) {
        if (tblent->key == key)
            return tblent;
    }
    return NULL;
}

static void *
PTABLE_fetch(PTABLE_t *tbl, const void *key)
{
    PTABLE_ENTRY_t const *const tblent = PTABLE_find(tbl, key);
    return tblent ? tblent->value : NULL;
}

/* add a new entry to a pointer => pointer table */

static void
PTABLE_store(PTABLE_t *tbl, void *key, void *value)
{
    PTABLE_ENTRY_t *tblent = PTABLE_find(tbl, key);

    if (tblent) {
        tblent->value = value;
    } else {
        const UV entry = PTABLE_HASH(key) & tbl->tbl_max;
        Newx(tblent, 1, PTABLE_ENTRY_t);

        tblent->key = key;
        tblent->value = value;
        tblent->next = tbl->tbl_ary[entry];
        tbl->tbl_ary[entry] = tblent;
        tbl->tbl_items++;
        if (tblent->next && (tbl->tbl_items > tbl->tbl_max))
            PTABLE_grow(tbl);
    }
}

/* double the hash bucket size of an existing ptr table */

static void
PTABLE_grow(PTABLE_t *tbl)
{
    PTABLE_ENTRY_t **ary = tbl->tbl_ary;
    const UV oldsize = tbl->tbl_max + 1;
    UV newsize = oldsize * 2;
    UV i;

    Renew(ary, newsize, PTABLE_ENTRY_t*);
    Zero(&ary[oldsize], newsize - oldsize, PTABLE_ENTRY_t*);
    tbl->tbl_max = --newsize;
    tbl->tbl_ary = ary;

    for (i = 0; i < oldsize; i++, ary++) {
        PTABLE_ENTRY_t **curentp, **entp, *ent;
        if (!*ary)
            continue;
        curentp = ary + oldsize;
        for (entp = ary, ent = *ary; ent; ent = *entp) {
            if ((newsize & PTABLE_HASH(ent->key)) != i) {
                *entp = ent->next;
                ent->next = *curentp;
                *curentp = ent;
                continue;
            } else {
                entp = &ent->next;
	    }
        }
    }
}

/* remove all the entries from a ptr table */

static void
PTABLE_clear(PTABLE_t *tbl)
{
    if (tbl && tbl->tbl_items) {
        register PTABLE_ENTRY_t * * const array = tbl->tbl_ary;
        UV riter = tbl->tbl_max;

        do {
            PTABLE_ENTRY_t *entry = array[riter];

            while (entry) {
                PTABLE_ENTRY_t * const oentry = entry;
                entry = entry->next;
                Safefree(oentry);
            }

            /* chocolateboy 2008-01-08
	     *
             * make sure we clear the array entry, so that subsequent probes fail
             */

            array[riter] = NULL;
        } while (riter--);

        tbl->tbl_items = 0;
    }
}

/* clear and free a ptr table */

static void
PTABLE_free(PTABLE_t *tbl)
{
    if (!tbl) {
        return;
    }
    PTABLE_clear(tbl);
    Safefree(tbl->tbl_ary);
    Safefree(tbl);
}


ptr_table.h
#include "ppport.h"

#if (PTRSIZE == 8)
#define PTR_TABLE_HASH(ptr) (PTR2UV(ptr) >> 3)
#else
#define PTR_TABLE_HASH(ptr) (PTR2UV(ptr) >> 2)
#endif

static PTR_TBL_t * ptr_table_new();
static void * ptr_table_fetch(PTR_TBL_t *tbl, void *sv);
static void ptr_table_store(PTR_TBL_t *tbl, void *oldv, void *newv);
static void ptr_table_grow(PTR_TBL_t *tbl);
static void ptr_table_clear(PTR_TBL_t *tbl);
static void ptr_table_free(PTR_TBL_t *tbl);

/* create a new pointer-mapping table */

static PTR_TBL_t *
ptr_table_new()
{
	PTR_TBL_t *tbl;
	Newz(0, tbl, 1, PTR_TBL_t);
	tbl->tbl_max = 511;
	tbl->tbl_items = 0;
	Newz(0, tbl->tbl_ary, tbl->tbl_max + 1, PTR_TBL_ENT_t*);
	return tbl;
}

/* map an existing pointer using a table */

static void *
ptr_table_fetch(PTR_TBL_t *tbl, void *sv)
{
	PTR_TBL_ENT_t *tblent;
	UV hash = PTR_TABLE_HASH(sv);
	tblent = tbl->tbl_ary[hash & tbl->tbl_max];

	for (; tblent; tblent = tblent->next) {
		if (tblent->oldval == sv) {
			/* Perl_warn(aTHX_ "    found value in ptr_table: 0x%x => 0x%x\n", sv, tblent->newval); */
			return tblent->newval;
		}
	}
	return (void*)NULL;
}

/* add a new entry to a pointer-mapping table */

static void
ptr_table_store(PTR_TBL_t *tbl, void *oldv, void *newv)
{
	PTR_TBL_ENT_t *tblent, **otblent;
	/* XXX this may be pessimal on platforms where pointers aren't good
	 * hash values e.g. if they grow faster in the most significant
	 * bits */
	UV hash = PTR_TABLE_HASH(oldv);
	bool empty = 1;
	/* Perl_warn(aTHX_ "    storing value in ptr_table: 0x%x => 0x%x\n", oldv, newv); */

	otblent = &tbl->tbl_ary[hash & tbl->tbl_max];
	for (tblent = *otblent; tblent; empty=0, tblent = tblent->next) {
		if (tblent->oldval == oldv) {
			tblent->newval = newv;
			return;
		}
	}
	Newz(0, tblent, 1, PTR_TBL_ENT_t);
	tblent->oldval = oldv;
	tblent->newval = newv;
	tblent->next = *otblent;
	*otblent = tblent;
	tbl->tbl_items++;
	if (!empty && tbl->tbl_items > tbl->tbl_max)
		ptr_table_grow(tbl);
}

/* double the hash bucket size of an existing ptr table */

static void
ptr_table_grow(PTR_TBL_t *tbl)
{
	PTR_TBL_ENT_t **ary = tbl->tbl_ary;
	UV oldsize = tbl->tbl_max + 1;
	UV newsize = oldsize * 2;
	UV i;

	Renew(ary, newsize, PTR_TBL_ENT_t*);
	Zero(&ary[oldsize], newsize-oldsize, PTR_TBL_ENT_t*);
	tbl->tbl_max = --newsize;
	tbl->tbl_ary = ary;
	for (i=0; i < oldsize; i++, ary++) {
		PTR_TBL_ENT_t **curentp, **entp, *ent;
		if (!*ary)
			continue;
		curentp = ary + oldsize;
		for (entp = ary, ent = *ary; ent; ent = *entp) {
			if ((newsize & PTR_TABLE_HASH(ent->oldval)) != i) {
				*entp = ent->next;
				ent->next = *curentp;
				*curentp = ent;
				continue;
			}
			else
				entp = &ent->next;
		}
	}
}

/* remove all the entries from a ptr table */

static void
ptr_table_clear(PTR_TBL_t *tbl)
{
	register PTR_TBL_ENT_t **array;
	register PTR_TBL_ENT_t *entry;
	register PTR_TBL_ENT_t *oentry = Null(PTR_TBL_ENT_t*);
	UV riter = 0;
	UV max;

	if (!tbl || !tbl->tbl_items) {
		return;
	}

	array = tbl->tbl_ary;
	entry = array[0];
	max = tbl->tbl_max;

	for (;;) {
		if (entry) {
			oentry = entry;
			entry = entry->next;
			Safefree(oentry);
		}
		if (!entry) {
			if (++riter > max) {
				break;
			}
			entry = array[riter];
		}
	}

	tbl->tbl_items = 0;
}

/* clear and free a ptr table */

static void
ptr_table_free(PTR_TBL_t *tbl)
{
	/* Perl_warn(aTHX_ "freeing ptr table\n"); */
	if (!tbl) {
		return;
	}
	ptr_table_clear(tbl);
	Safefree(tbl->tbl_ary);
	Safefree(tbl);
}


lib/autobox.pm
package autobox;

use 5.008;

use strict;
use warnings;

use Carp;
use XSLoader;
use Scalar::Util;
use Scope::Guard;
use Storable;

# XXX this declaration must be on a single line
# https://metacpan.org/pod/version#How-to-declare()-a-dotted-decimal-version
use version 0.77; our $VERSION = version->declare('v3.0.1');

XSLoader::load 'autobox', $VERSION;

use autobox::universal (); # don't import

############################################# PRIVATE ###############################################

my $SEQ            = 0;  # unique identifier for synthetic classes
my $BINDINGS_CACHE = {}; # hold a reference to the bindings hashes
my $CLASS_CACHE    = {}; # reuse the same synthetic class if the @isa has been seen before

# all supported types
# the boolean indicates whether the type is a real internal type (as opposed to a virtual type)
my %TYPES = (
    UNDEF     => 1,
    INTEGER   => 1,
    FLOAT     => 1,
    NUMBER    => 0,
    STRING    => 1,
    SCALAR    => 0,
    ARRAY     => 1,
    HASH      => 1,
    CODE      => 1,
    UNIVERSAL => 0
);

# type hierarchy: keys are parents, values are (depth, children) pairs
my %ISA = (
    UNIVERSAL => [ 0, [ qw(SCALAR ARRAY HASH CODE) ] ],
    SCALAR    => [ 1, [ qw(STRING NUMBER) ] ],
    NUMBER    => [ 2, [ qw(INTEGER FLOAT) ] ]
);

# default bindings when no args are supplied
my %DEFAULT = (
    SCALAR => 'SCALAR',
    ARRAY  => 'ARRAY',
    HASH   => 'HASH',
    CODE   => 'CODE'
);

# reinvent List::MoreUtils::uniq to keep the dependencies light - return a reference
# to an array containing (in order) the unique members of the supplied list
sub _uniq($) {
    my $list = shift;
    my (%seen, @uniq);

    for my $element (@$list) {
        next if ($seen{$element});
        push @uniq, $element;
        $seen{$element} = 1;
    }

    return [ @uniq ];
}

# create a shim class - actual methods are implemented by the classes in its @ISA
#
# as an optimization, return the previously-generated class
# if we've seen the same (canonicalized) @isa before
sub _generate_class($) {
    my $isa = _uniq(shift);

    # As an optimization, simply return the class if there's only one.
    # This speeds up method lookup as the method can (often) be found directly in the stash
    # rather than in the ISA hierarchy with its attendant AUTOLOAD-related overhead
    if (@$isa == 1) {
        my $class = $isa->[0];
        _make_class_accessor($class); # NOP if it has already been added
        return $class;
    }

    my $key = Storable::freeze($isa);

    return $CLASS_CACHE->{$key} ||= do {
        my $class = sprintf('autobox::_shim_%d_', ++$SEQ);
        my $synthetic_class_isa = _get_isa($class); # i.e. autovivify

        @$synthetic_class_isa = @$isa;
        _make_class_accessor($class);
        $class;
    };
}

# expose the autobox class (for can, isa etc.)
# https://rt.cpan.org/Ticket/Display.html?id=55565
sub _make_class_accessor ($) {
    my $class = shift;
    return unless (defined $class);

    {
        no strict 'refs';
        *{"$class\::autobox_class"} = sub { $class } unless (*{"$class\::autobox_class"}{CODE});
    }
}

# pretty-print the bindings hash by showing its values as the inherited classes rather than the synthetic class
sub _pretty_print($) {
    my $hash = { %{ shift() } }; # clone the hash to isolate it from the original

    # reverse() turns a hash that maps an isa signature to a class name into a hash that maps
    # a class name into a boolean
    my %synthetic = reverse(%$CLASS_CACHE);

    for my $type (keys %$hash) {
        my $class = $hash->{$type};
        $hash->{$type} = $synthetic{$class} ? [ _get_isa($class) ] : [ $class ];
    }

    return $hash;
}

# default sub called when the DEBUG option is supplied with a true value
# prints the assigned bindings for the current scope
sub _debug ($) {
    my $bindings = shift;
    require Data::Dumper;
    no warnings qw(once);
    local ($|, $Data::Dumper::Indent, $Data::Dumper::Terse, $Data::Dumper::Sortkeys) = (1, 1, 1, 1);
    print STDERR Data::Dumper::Dumper($bindings), $/;
}

# return true if $ref ISA $class - works with non-references, unblessed references and objects
# we can't use UNIVERSAL::isa to test if a value is an array ref;
# if the value is 'ARRAY', and that package exists, then UNIVERSAL::isa('ARRAY', 'ARRAY') is true!
sub _isa($$) {
    my ($ref, $class) = @_;
    return Scalar::Util::blessed($ref) ? $ref->isa($class) : ref($ref) eq $class;
}

# get/autovivify the @ISA for the specified class
sub _get_isa($) {
    my $class = shift;
    my $isa   = do {
        no strict 'refs';
        *{"$class\::ISA"}{ARRAY};
    };
    return wantarray ? @$isa : $isa;
}

# install a new set of bindings for the current scope
#
# XXX this could be refined to reuse the same hashref if its contents have already been seen,
# but that requires each (frozen) hash to be cached; at best, it may not be much of a win, and at
# worst it will increase bloat
sub _install ($) {
    my $bindings = shift;
    $^H{autobox} = $bindings;
    $BINDINGS_CACHE->{$bindings} = $bindings; # keep the $bindings hash alive
}

# return the supplied class name or a new class name made by appending the specified
# type to the namespace prefix
sub _expand_namespace($$) {
    my ($class, $type) = @_;

    # make sure we can weed out classes that are empty strings or undef by returning an empty list
    Carp::confess("_expand_namespace not called in list context") unless (wantarray);

    if ((defined $class) && ($class ne '')) {
        ($class =~ /::$/) ? "$class$type" : $class;
    } else { # return an empty list
        ()
    }
}

############################################# PUBLIC (Methods) ###############################################

# enable some flavour of autoboxing in the current scope
sub import {
    my $class = shift;
    my %args = ((@_ == 1) && _isa($_[0], 'HASH')) ? %{shift()} : @_; # hash or hashref
    my $debug = delete $args{DEBUG};

    %args = %DEFAULT unless (%args); # wait till DEBUG has been deleted

    # normalize %args so that it has a (possibly empty) array ref for all types, both real and virtual
    for my $type (keys %TYPES) {
        if (exists $args{$type}) { # exists() as the value may be undef (or ''), meaning "don't default this type"
            if (_isa($args{$type}, 'ARRAY')) {
                $args{$type} = [ @{$args{$type}} ]; # clone the array ref to isolate changes
            } else {
                $args{$type} = [ $args{$type} ];
            }
        } else {
            $args{$type} = [];
        }
    }

    # if supplied, fill in defaults for unspecified SCALAR, ARRAY, HASH and CODE bindings
    # must be done before the virtual type expansion below as one of the defaults, SCALAR, is a
    # virtual type

    my $default = delete $args{DEFAULT};

    if ($default) {
        $default = [ $default ] unless (_isa($default, 'ARRAY')); # no need to clone as we flatten it each time

        for my $type (keys %DEFAULT) {
            # don't default if a binding has already been supplied; this may include an undef value meaning
            # "don't default this type" e.g.
            #
            #     use autobox
            #         DEFAULT => 'MyDefault',
            #         HASH    => undef;
            #
            # undefs are winnowed out by _expand_namespace

            next if (@{$args{$type}});
            push @{$args{$type}}, map { _expand_namespace($_, $type) } @$default;
        }
    }

    # expand the virtual type "macros" from the root to the leaves
    for my $vtype (sort { $ISA{$a}->[0] <=> $ISA{$b}->[0] } keys %ISA) {
        next unless ($args{$vtype});

        my @types = @{$ISA{$vtype}->[1]};

        for my $type (@types) {
            if (_isa($args{$vtype}, 'ARRAY')) {
                push @{$args{$type}}, map { _expand_namespace($_, $vtype) } @{$args{$vtype}};
            } else {
                # _expand_namespace returns an empty list if $args{$vtype} is undef (or '')
                push @{$args{$type}}, _expand_namespace($args{$vtype}, $vtype);
            }
        }

        delete $args{$vtype};
    }

    my $bindings; # custom typemap

    # clone the bindings hash if available
    #
    # we may be assigning to it, and we don't want to contaminate outer/previous bindings
    # with nested/new bindings
    #
    # as of 5.10, references in %^H get stringified at runtime, but we don't need them then

    $bindings = $^H{autobox} ? { %{ $^H{autobox} } } : {};

    # sanity check %args, expand the namespace prefixes into class names,
    # and copy values to the $bindings hash

    my %synthetic = reverse(%$CLASS_CACHE); # synthetic class name => bool - see _pretty_print

    for my $type (keys %args) {
        # we've handled the virtual types, so we only need to check that this is a valid (real) type
        Carp::confess("unrecognized option: '", (defined $type ? $type : '<undef>'), "'") unless ($TYPES{$type});

        my (@isa, $class);

        if ($class = $bindings->{$type}) {
            @isa = $synthetic{$class} ? _get_isa($class) : ($class);
        }

        # perform namespace expansion; dups are removed in _generate_class below
        push @isa, map { _expand_namespace($_, $type) } @{$args{$type}};

        $bindings->{$type} = [ @isa ]; # assign the (possibly) new @isa for this type
    }

    # replace each array ref of classes with the name of the generated class.
    # if there's only one class in the type's @ISA (e.g. SCALAR => 'MyScalar') then
    # that class is used; otherwise a shim class whose @ISA contains the two or more classes
    # is created

    for my $type (keys %$bindings) {
        my $isa = $bindings->{$type};

        # delete empty arrays e.g. use autobox SCALAR => []
        if (@$isa == 0) {
            delete $bindings->{$type};
        } else {
            # associate the synthetic/single class with the specified type
            $bindings->{$type} = _generate_class($isa);
        }
    }

    # This turns on autoboxing i.e. the method call checker sets a flag on the method call op
    # and replaces its default handler with the autobox implementation.
    #
    # It needs to be set unconditionally because it may have been unset in unimport

    $^H |= 0x80020000; # set HINT_LOCALIZE_HH + an unused bit to work around a %^H bug

    # install the specified bindings in the current scope
    _install($bindings);

    # this is %^H as an integer - it changes as scopes are entered/exited
    # we don't need to stack/unstack it in %^H as %^H itself takes care of that
    # note: we need to call this *after* %^H is referenced (and possibly created) above

    my $scope = _scope();
    my $old_scope = exists($^H{autobox_scope})? $^H{autobox_scope} : 0;
    my $new_scope; # is this a new (top-level or nested) scope?

    if ($scope == $old_scope) {
        $new_scope = 0;
    } else {
        $^H{autobox_scope} = $scope;
        $new_scope = 1;
    }

    # warn "OLD ($old_scope) => NEW ($scope): $new_scope ", join(':', (caller(1))[0 .. 2]), $/;

    if ($debug) {
        $debug = \&_debug unless (_isa($debug, 'CODE'));
        $debug->(_pretty_print($bindings));
    }

    return unless ($new_scope);

    # This sub is called when this scope's $^H{autobox_leave} is deleted, usually when
    # %^H is destroyed at the end of the scope, but possibly directly in unimport()
    #
    # _enter splices in the autobox method call checker and method call op
    # if they're not already enabled
    #
    # _leave performs the necessary housekeeping to ensure that the default
    # checker and op are restored when autobox is no longer in scope

    my $guard = Scope::Guard->new(sub { _leave() });
    $^H{autobox_leave} = $guard;

    _enter();
}

# delete one or more bindings; if none remain, disable autobox in the current scope
#
# note: if bindings remain, we need to create a new hash (initially a clone of the current
# hash) so that the previous hash (if any) is not contaminated by new deletion(s)
#
#   use autobox;
#
#       "foo"->bar;
#
#   no autobox qw(SCALAR); # don't clobber the default bindings for "foo"->bar
#
# however, if there are no more bindings we can remove all traces of autobox from the
# current scope.

sub unimport {
    my ($class, @args) = @_;

    # the only situation in which there is no bindings hash is if this is a "no autobox"
    # that precedes any "use autobox", in which case we don't need to turn autoboxing off as it's
    # not yet been turned on
    return unless ($^H{autobox});

    my $bindings;

    if (@args) {
        $bindings = { %{$^H{autobox}} }; # clone the current bindings hash
        my %args = map { $_ => 1 } @args;

        # expand any virtual type "macros"
        for my $vtype (sort { $ISA{$a}->[0] <=> $ISA{$b}->[0] } keys %ISA) {
            next unless ($args{$vtype});

            # we could delete the types directly from $bindings here, but we may as well pipe them
            # through the option checker below to ensure correctness
            $args{$_} = 1 for (@{$ISA{$vtype}->[1]});

            delete $args{$vtype};
        }

        for my $type (keys %args) {
            # we've handled the virtual types, so we only need to check that this is a valid (real) type
            Carp::confess("unrecognized option: '", (defined $type ? $type : '<undef>'), "'") unless ($TYPES{$type});
            delete $bindings->{$type};
        }
    } else { # turn off autoboxing
        $bindings = {}; # empty hash to trigger full deletion below
    }

    if (%$bindings) {
        _install($bindings);
    } else { # remove all traces of autobox from the current scope
        $^H &= ~0x80020000; # unset HINT_LOCALIZE_HH + the additional bit
        delete $^H{autobox};
        delete $^H{autobox_scope};
        delete $^H{autobox_leave}; # triggers the leave handler
    }
}

1;


Makefile.PL
use 5.008;

use strict;
use warnings;

use ExtUtils::MakeMaker;
use Config;

my $EUMM_VERSION = eval($ExtUtils::MakeMaker::VERSION);
my $OPTIMIZE;

if ($Config{gccversion}) {
    $OPTIMIZE = '-O3 -Wall';
    $OPTIMIZE .= ' -g -Wextra -Wdeclaration-after-statement' if (-d 'dev');
} elsif ($Config{osname} eq 'MSWin32') {
    $OPTIMIZE = '-O2 -W4';
} else {
    $OPTIMIZE = $Config{optimize};
}

my $META_MERGE = {
    resources => {
        repository => 'https://github.com/chocolateboy/autobox',
        bugtracker => 'https://github.com/chocolateboy/autobox/issues',
    },
};

my $TEST_REQUIRES = {
    'IPC::System::Simple'   => '1.25',
    'Test::Fatal'           => '0.014',
    'Parallel::ForkManager' => '2.02',
};

WriteMakefile(
    NAME         => 'autobox',
    VERSION_FROM => 'lib/autobox.pm',

    # compatibility in case module was previously installed to lib
    INSTALLDIRS => ($] >= 5.011 ? 'site' : 'perl'),

    PREREQ_PM => {
        'Scope::Guard' => '0.21',
        'version'      => '0.77',
    },

    ABSTRACT_FROM => 'lib/autobox.pod',
    AUTHOR        => 'chocolateboy <chocolate@cpan.org>',
    INC           => '-I.',
    OPTIMIZE      => $OPTIMIZE,
    ($EUMM_VERSION >= 6.5503 ? (BUILD_REQUIRES   => $TEST_REQUIRES) : ()),
    ($EUMM_VERSION >= 6.31   ? (LICENSE          => 'artistic_2')   : ()),
    ($EUMM_VERSION >= 6.46   ? (META_MERGE       => $META_MERGE)    : ()),
    ($EUMM_VERSION >= 6.48   ? (MIN_PERL_VERSION => '5.8.0')        : ()),
);
